src/WebJobs.Script.WebHost/Diagnostics/SystemLogger.cs

```diff
@@ -51,75 +75,141 @@ public bool IsEnabled(LogLevel logLevel)
         {
-            // propagate special exceptions through the EventManager
-            var stateProps = state as IEnumerable<KeyValuePair<string, object>> ?? new Dictionary<string, object>();
+            if (!IsEnabled(logLevel) || IsUserLog(state))
+            {
+                return;
+            }
 
-            string source = _categoryName ?? Utility.GetStateValueOrDefault<string>(stateProps, ScriptConstants.LogPropertySourceKey);
-            if (exception is FunctionIndexingException && _eventManager != null)
+            var logItem = new LogItem
+            {
+                LogLevel = logLevel,
+                EventId = eventId,
+                State = state,
+                Exception = exception
+            };
+
+            if (formatter != null)
             {
-                _eventManager.Publish(new FunctionIndexingEvent("FunctionIndexingException", source, exception));
+                logItem.Formatter = (s, e) =>
+                {
+                    return formatter.Invoke(state, exception);
+                };
             }
 
-            // User logs are not logged to system logs.
-            if (!IsEnabled(logLevel) || IsUserLog(state))
+            lock (_syncLock)
             {
-                return;
+                _logItemQueue.Add(logItem);
             }
+        }
 
-            string formattedMessage = formatter?.Invoke(state, exception);
+        private void LogCore(LogItem logItem)
+        {
+            // propagate special exceptions through the EventManager
+            var stateProps = logItem.State as IEnumerable<KeyValuePair<string, object>>;
+            string source = _categoryName ?? Utility.GetStateValueOrDefault<string>(stateProps, ScriptConstants.LogPropertySourceKey);
+            if (logItem.Exception is FunctionIndexingException && _eventManager != null)
+            {
+                _eventManager.Publish(new FunctionIndexingEvent(nameof(FunctionIndexingException), source, logItem.Exception));
+            }
 
             // If we don't have a message, there's nothing to log.
+            string formattedMessage = logItem.Formatter?.Invoke(logItem.State, logItem.Exception);
             if (string.IsNullOrEmpty(formattedMessage))
             {
                 return;
             }
 
-            IDictionary<string, object> scopeProps = _scopeProvider.GetScopeDictionary();
-
             // Apply standard event properties
             // Note: we must be sure to default any null values to empty string
             // otherwise the ETW event will fail to be persisted (silently)
-            string subscriptionId = _environment.GetSubscriptionId() ?? string.Empty;
-            string appName = _environment.GetAzureWebsiteUniqueSlotName() ?? string.Empty;
-            string summary = Sanitizer.Sanitize(formattedMessage) ?? string.Empty;
```

Comment:
As you can see, we're doing way too much stuff on each log. One huge item was this Sanitizer.Sanitize operation, which was doing a bunch of string search/replace. We don't need to pay the overhead of that on every single log statement. Instead, the caller should know whether the log might contain sensitive info, and sanitize itself. I'm actually not aware of any logs where this is needed anymore.

Reply:
I think this was specifically added to remove secrets if they are coming from exceptions. Probably safe to remove. tagging @brettsam to confirm.

Reply:
@pragnagopa is right. We cannot remove the sanitizing as we don't know what logs are being written from exceptions. Perhaps we can come back and improve this with some perf benchmarks, but we cannot remove this.

Reply:
I was trying to think whether we could check whether the passed-in exception was null here and only sanitize then, but I worry that somewhere else is calling `ToString()` on a user exception before logging it.

Another approach is to look at this `Sanitize()` method, which looks pretty inefficient (I copied it from internal Antares code back when we initially had the fire drill that required this). We should be able to micro-benchmark that method itself and get it much faster. Is that something we could follow up with so it doesn't block this getting in?